sharedbuffers/mapped_struct.py

```diff
@@ -297,6 +306,226 @@ def unpack_from(cls, buf, offs, idmap = None, array = array.array, itemsizes = {
             raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
         return rv
 
+@cython.cclass
+class proxied_list(object):
+    cython.declare(
+        buf = object,
+        pybuf = 'Py_buffer',
+        offs = cython.ulonglong,
+        dataoffs = cython.ulonglong,
+        itemsize = cython.ulonglong,
+        objlen = cython.ulonglong,
+        index = 'const unsigned long *',
+        dcode = object,
+        idmap = dict,
+        struct = object,
+    )
+
+    if cython.compiled:
+        def __del__(self):
+            if self.pybuf.buf != cython.NULL:
+                PyBuffer_Release(cython.address(self.pybuf))  # lint:ok
+
+    def __init__(self, buf, offs, dataoffs, dcode, itemsize, objlen, idmap):
+        self.dcode = dcode
+        self.itemsize = itemsize
+        self.idmap = idmap
+        self.offs = offs
+        self.buf = buf
+        self.objlen = objlen
+        self.dataoffs = dataoffs
+
+        if cython.compiled:
+            self._cinit()
+        else:
+            if dcode in ('B','b','H','h','I','i', 'd'):
+                self.struct = struct.Struct(dcode)
+            elif dcode == 'q':
+                self.struct = struct.Struct('l')
+            elif dcode == 't':
+                self.struct = struct.Struct('l')
+            else:
+                raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
+
+    @cython.ccall
+    def _cinit(self):
+        self.pybuf.buf = cython.NULL
+        PyObject_GetBuffer(self.buf, cython.address(self.pybuf), PyBUF_SIMPLE)  # lint:ok
+        if self.dcode == 't':
+            self.index = cython.cast(cython.p_ulong, cython.cast(cython.p_uchar, self.pybuf.buf) + self.dataoffs)
+        else:
+            self.index = cython.NULL
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        # Same format as tuple, only different base type
+        return mapped_tuple.pack_into(obj, buf, offs, idmap, implicit_offs)
+
+    @classmethod
+    def unpack_from(cls, buf, offs, idmap = None, array = array.array, itemsizes = {
+                dtype : array.array(dtype, []).itemsize
+                for dtype in ('B','b','H','h','I','i','l','d')
+            } ):
+        if idmap is None:
+            idmap = {}
+        if offs in idmap:
+            return idmap[offs]
+
+        buf = _likerobuffer(buf)
+        dcode = buf[offs]
+        dataoffs = offs
+
+        if dcode in ('B','b','H','h','I','i'):
```

Comment:
Check the code generated here.

Ideally, you'll want a switch. You may have to use `@cython.use_switch(True)` and make sure the `dcode` is a char.